### ISSUE TITLE
Advancing code relations

### ISSUE THREAD

**Issue by AndrzejWawa** [ccbogel/QualCoder#685]
Hi,

I think that code relations report could be advanced. Already, there can be generated detailed and flexible table which can be exported (and this should kept), however, to utilise it, user needs to know spreadsheet software or SQL, sometimes at quite good level. So, some additional options would be very appreciated, especially, that code relations is one of the most important report in CAQDAS, as well as additional options have been introduced by other software, e.g. Maxqda: https://www.maxqda.com/help-mx20/visual-tools/code-relations-browser-visualizing-overlapping-codes.

My propositions:

1) In general, I would introduce new functions at the top of the repot, as presented on the image:

![Code relations](https://user-images.githubusercontent.com/77200214/178121311-95703efd-24f4-419d-8e7e-873340fe1a28.png)

It is important, that when you select your options and generate table, the selected options are still visible (as, in the case of selecting coders in coding comparison).

2) First of all, there could be option to show only selected relations (expressd in the full words, e.g. "Exact everlapping (E)"), so 1) All, 2) exact overlapping, 3) not-exact overlapping, 4) inclusion, 5) proximity), 6) combinations, e.g. exact overlapping and not-exact overlapping. For this, user could have possibility to select more than one relation,

3) Then, possibility to select file or files. All files by default.

4) Then, more complicated, but important issue, to be able to select option to show relations with ALL selected codes. Now, if you select codes, QC shows all relations in which at least ONE of selected codes appears. So, it is rather logic CodeA AND CodeB AND CodeC. However, it is often very needed to show only relations where ALL selected codes are present, so, the logic CodeA AND CodeB AND CodeC. For example, if you select option "Show only relations with ALL selected codes", and you select Code A,B,C and relation "E", QC will show you ONLY segments which are coded SIMULTANEOUSLY by ALL three codes. It is more complicated by when it comes to the other relations, but it can be based on the SQL we discussed:

SELECT code_relations.Fid , Code0_name, Code0, Code0_pos0, Code0_pos1, Code1_name, Code1, Code1_pos0, Code1_pos1, Relation, ctid0, ctid1, text0, Text_overlap, text1, 
ctcode0.memo as "memo0", code_text.memo as "memo1"
FROM code_relations left join code_text on code_text.ctid = ctid1 -- CHANGE TO ctid = ctidCode1 to show memos for code number3
left join code_text as ctcode0 on ctcode0.ctid=ctid0

WHERE
Relation = "O" --OPTION1:  SLECT KIND OR KINDS OF RELATION(S)

--AND Fid = 1  --OPTION 2: SELECT FILE OR FILES

AND                 -- OPTION3: SELECT CODES (CHANGE NUMBER OF CODES IN ALL "Code0 =" and "Code1 =")
(
ctid0 IN 
	(SELECT ctid0 FROM code_relations WHERE Code0 = 3 AND Code1 = 2 AND Relation= "O" AND ctid1 IN -- 3-2
 		(SELECT ctid0 FROM code_relations WHERE Code0= 2 AND Code1 = 1 AND Relation= "O" AND ctid1 IN -- 2-1 
   			 (SELECT ctid1 FROM code_relations WHERE Code0=3 AND Code1 = 1 AND Relation = "O")) -- 1-3

AND ctid0 IN 
	(SELECT ctid0 FROM code_relations WHERE Code0= 3 AND Code1 = 1 AND Relation= "O" AND ctid1 IN -- 3-1
 		(SELECT ctid1 FROM code_relations WHERE Code1 = 1 AND Code0 = 2 AND Relation = "O" AND ctid0 IN -- 1-2
    			  (SELECT ctid1 FROM code_relations WHERE Code1= 2 AND Code0 = 3 AND Relation = "O") -- 2-3
    )
  )
)

-- TO SHOW ALSO RELATION BETWEEN CODE 2 AND CODE 1 
OR ctid0 IN 
	(SELECT ctid0 FROM code_relations WHERE Code0 = 2 AND Code1 = 1 AND Relation= "O" AND ctid1 IN -- 2-1
 		 (SELECT ctid1 FROM code_relations WHERE Code0= 3 AND Code1 = 1 AND Relation= "O" AND ctid0 IN -- 1-3 
  			  (SELECT ctid0 FROM code_relations WHERE Code0=3 AND Code1 = 2 AND Relation = "O")) -- 3-2

AND ctid0 IN 
 	 (SELECT ctid1 FROM code_relations WHERE Code0= 3 AND Code1 = 2 AND Relation= "O" AND ctid0 IN -- 2-3
    			(SELECT ctid0 FROM code_relations WHERE Code1 = 1 AND Code0 = 3 AND Relation = "O" AND ctid1 IN -- 3-1
     				 (SELECT ctid1 FROM code_relations WHERE Code1= 1 AND Code0 = 2 AND Relation = "O") -- 1-2
    )
  )
  )
)

ORDER BY ctid0 ASC -- OPTION 4: SORT BY (ASC, DESC)

-- GROUP BY ctid0   -- OPTION 5: SHOW LESS ROWS (also allow for showing only one row for each comparison)

5) The SQL above can be inserted additionally to default queries in SQL report, for the case some users would like to suit it to their research.

6) Option show less rows could use SQL function GROUP BY. For example, in the case of E relation between 3 codes, it can group this code by the pos0, so, despite 3 rows it will display only one row. Again, it is more complicated in the case of the other relations, but there, rows can be grouped by ctid0. so, 2 rows will be shown instead of 3.

7) Number of relations with ALL codes will show you aggregated relations, so, in the case of E relations between A, B, and C, it will show you number of relations, previously GROUPED BY pos0. In other words, the number of cases, where the one segment is coded simultaneously by A, B, AND C/

8) Avarage distance statistics will be shown in the case of P relation and how average distance between selected code. For example, if you select A, B, C, and P relations, there will be information on the average relation between A and B, B and C, and A and C.

9) I've changed my mind, It is good as it is. But, traditionally, search box could be added also for this report, as searching is very often needed.

10) AND NOT option will be optionally and related to show relations with ALL selected codes. So, for example, you can show only relations E with codes A, B, AND C, AND NOT D (SQL logical operator).

11) The memo0 and memo1 could be added to the table displayed in QC, as some users may analyse mainly memos summarising codings, not the codings directly.

12) Moreover, the righ-click menu options (now available in SQL statements) could be added (e.g. sorting, filtering) By default table should be sorted ascending by ctid1, as it will support interpretation (ctid's are generated in time order, as I understand).

13) And the cherry on top - righ-clicking option for "number of relations with ALL codes" and "average distance" to generate bar charts with plotly. I think that this will make great impression! :)

As I learn SQL a bit, I can support you when I will be able.

**Comment by ccbogel** [comment]
OK. I have added selections for the relationship type and for file selection.

**Comment by ccbogel** [comment]
@AndrzejWawa The sql statements:
AND Relation= "O" 
Everywhere change these to:
AND Relation in ("O", "I", "E")

Because you will want to see the codes that are entirely surrounded by another code "I" and the codes that have an exact match "E"

**Comment by AndrzejWawa** [comment]
Hmm, I meant that relation types will be selected conventionally, through the drop-down list you introduced. What I suggest it is an introduction of additional option to select/unselect by users, which changes the calculation method.

For example. let's say that we have code A, B, and C. And you selecting these three codes and select "E" relation. The result is:

NOW:

- segments coded by A and B and C,
- by A and B,
- by B and C,
- by A and C.

IN PROPOSED SQL:
- ONLY segments coded by A and B and C.

So. there is need for option to showing only segments which are at the same time coded by ALL selected codes. If you have codes "dogs", "happy", "from Poland", very often you would seek only for segments about dogs which are at the same time 1) dogs, 2) happy, 3) and from Poland. However, current calculation method does not allow for this, because it will show you also only happy dogs, but not from Poland, or dogs from Poland, but not happy.


SQL above was written for the case, that user selects "O"relationship, however, here are versions for "E" and "I":

SELECT code_relations.Fid , Code0_name, Code0, Code0_pos0, Code0_pos1, Code1_name, Code1, Code1_pos0, Code1_pos1, Relation, ctid0, ctid1, text0, Text_overlap, text1, 
ctcode0.memo as "memo0", code_text.memo as "memo1"
FROM code_relations left join code_text on code_text.ctid = ctid1 -- CHANGE TO ctid = ctidCode1 to show memos for code number3
left join code_text as ctcode0 on ctcode0.ctid=ctid0

WHERE
Relation = "E" --OPTION1:  SLECT KIND OR KINDS OF RELATION(S)

--AND Fid = 1  --OPTION 2: SELECT FILE OR FILES

AND                 -- OPTION3: SELECT CODES (CHANGE NUMBER OF CODES IN ALL "Code0 =" and "Code1 =")
(
ctid0 IN 
	(SELECT ctid0 FROM code_relations WHERE Code0 = 3 AND Code1 = 2 AND Relation= "E" AND ctid1 IN -- 3-2
 		(SELECT ctid0 FROM code_relations WHERE Code0= 2 AND Code1 = 1 AND Relation= "E" AND ctid1 IN -- 2-1 
   			 (SELECT ctid1 FROM code_relations WHERE Code0=3 AND Code1 = 1 AND Relation = "E")) -- 1-3

AND ctid0 IN 
	(SELECT ctid0 FROM code_relations WHERE Code0= 3 AND Code1 = 1 AND Relation= "E" AND ctid1 IN -- 3-1
 		(SELECT ctid1 FROM code_relations WHERE Code1 = 1 AND Code0 = 2 AND Relation = "E" AND ctid0 IN -- 1-2
    			  (SELECT ctid1 FROM code_relations WHERE Code1= 2 AND Code0 = 3 AND Relation = "E") -- 2-3
    )
  )
)

-- TO SHOW ALSO RELATION BETWEEN CODE 2 AND CODE 1 
OR ctid0 IN 
	(SELECT ctid0 FROM code_relations WHERE Code0 = 2 AND Code1 = 1 AND Relation= "E" AND ctid1 IN -- 2-1
 		 (SELECT ctid1 FROM code_relations WHERE Code0= 3 AND Code1 = 1 AND Relation= "E" AND ctid0 IN -- 1-3 
  			  (SELECT ctid0 FROM code_relations WHERE Code0=3 AND Code1 = 2 AND Relation = "E")) -- 3-2

AND ctid0 IN 
 	 (SELECT ctid1 FROM code_relations WHERE Code0= 3 AND Code1 = 2 AND Relation= "E" AND ctid0 IN -- 2-3
    			(SELECT ctid0 FROM code_relations WHERE Code1 = 1 AND Code0 = 3 AND Relation = "E" AND ctid1 IN -- 3-1
     				 (SELECT ctid1 FROM code_relations WHERE Code1= 1 AND Code0 = 2 AND Relation = "E") -- 1-2
    )
  )
  )
)

ORDER BY ctid0 ASC -- OPTION 4: SORT BY (ASC, DESC)

-- GROUP BY ctid0   -- OPTION 5: SHOW LESS ROWS (also allow for showing only one row for each comparison)




SELECT code_relations.Fid , Code0_name, Code0, Code0_pos0, Code0_pos1, Code1_name, Code1, Code1_pos0, Code1_pos1, Relation, ctid0, ctid1, text0, Text_overlap, text1, 
ctcode0.memo as "memo0", code_text.memo as "memo1"
FROM code_relations left join code_text on code_text.ctid = ctid1 -- CHANGE TO ctid = ctidCode1 to show memos for code number3
left join code_text as ctcode0 on ctcode0.ctid=ctid0

WHERE
Relation = "I" --OPTION1:  SLECT KIND OR KINDS OF RELATION(S)

--AND Fid = 1  --OPTION 2: SELECT FILE OR FILES

AND                 -- OPTION3: SELECT CODES (CHANGE NUMBER OF CODES IN ALL "Code0 =" and "Code1 =")
(
ctid0 IN 
	(SELECT ctid0 FROM code_relations WHERE Code0 = 3 AND Code1 = 2 AND Relation= "I" AND ctid1 IN -- 3-2
 		(SELECT ctid0 FROM code_relations WHERE Code0= 2 AND Code1 = 1 AND Relation= "I" AND ctid1 IN -- 2-1 
   			 (SELECT ctid1 FROM code_relations WHERE Code0=3 AND Code1 = 1 AND Relation = "I")) -- 1-3

AND ctid0 IN 
	(SELECT ctid0 FROM code_relations WHERE Code0= 3 AND Code1 = 1 AND Relation= "I" AND ctid1 IN -- 3-1
 		(SELECT ctid1 FROM code_relations WHERE Code1 = 1 AND Code0 = 2 AND Relation = "I" AND ctid0 IN -- 1-2
    			  (SELECT ctid1 FROM code_relations WHERE Code1= 2 AND Code0 = 3 AND Relation = "I") -- 2-3
    )
  )
)

-- TO SHOW ALSO RELATION BETWEEN CODE 2 AND CODE 1 
OR ctid0 IN 
	(SELECT ctid0 FROM code_relations WHERE Code0 = 2 AND Code1 = 1 AND Relation= "I" AND ctid1 IN -- 2-1
 		 (SELECT ctid1 FROM code_relations WHERE Code0= 3 AND Code1 = 1 AND Relation= "I" AND ctid0 IN -- 1-3 
  			  (SELECT ctid0 FROM code_relations WHERE Code0=3 AND Code1 = 2 AND Relation = "I")) -- 3-2

AND ctid0 IN 
 	 (SELECT ctid1 FROM code_relations WHERE Code0= 3 AND Code1 = 2 AND Relation= "I" AND ctid0 IN -- 2-3
    			(SELECT ctid0 FROM code_relations WHERE Code1 = 1 AND Code0 = 3 AND Relation = "I" AND ctid1 IN -- 3-1
     				 (SELECT ctid1 FROM code_relations WHERE Code1= 1 AND Code0 = 2 AND Relation = "I") -- 1-2
    )
  )
  )
)

ORDER BY ctid0 ASC -- OPTION 4: SORT BY (ASC, DESC)

-- GROUP BY ctid0   -- OPTION 5: SHOW LESS ROWS (also allow for showing only one row for each comparison)

**Comment by ccbogel** [comment]
I am moving the 3 codes SQL to a different thread. Its becoming out of place and overwhelming here.

**Comment by ccbogel** [comment]
Can you explain further some of the other options you added to the picture.
One is average distance - several issues here:
1 is mean(average) the best statistic - probably not - I suggest median and range, or IQR.
2 how would this be displayed as there are going to be multiple combinations of CodeA to CodeB,C,D ...  A separate window?

**Comment by AndrzejWawa** [comment]
Hmm, maybe it can work as follows: firstly you select codes and (optionally) files, and then, select proximity statistics (you can provide all options in the form of drop-down list - as in the case of relation types). Then, after clicking run button, there will be generated table with names of codes as columns headings (e.g. "name1-name2", "name2-name3"), and statistics types as names of rowa ("median", "range", "mean", "IQR").

In other words, when you select codes and type of relation, there is generated relational table, when you select codes and type of statistics, there is generated table with statistics. And both types can be generated also for only selected files and exported through CSV.

**Comment by AndrzejWawa** [comment]
Good work!

Can you add tooltip to show the whole name of kindd of relationship? They are not visible.

![obraz](https://user-images.githubusercontent.com/77200214/179257066-e00c3dd5-8516-448a-9ac8-f00fbdb7dce7.png)
![obraz](https://user-images.githubusercontent.com/77200214/179257066-e00c3dd5-8516-448a-9ac8-f00fbdb7dce7.png)

**Comment by ccbogel** [comment]
Done

**Comment by ccbogel** [comment]
@AndrzejWawa I will close this issue now, but can you open new issues (from your picture) with more description of what you want. e.g. show less rows, AND NOT
